File: Data Structures & Algorithms/decode-ways/submission-1.java
 class Solution {
          public int numDecodings(String s) {
                   int[] dp = new int[s.length()];
                            // 1 ~ 26
                             
                                      return dfs(s, 0, dp);
                                           }
                                            
                                                 int dfs(String s, int idx, int[] dp) {
                                                          if (idx == s.length()) return 1;
                                                                   if (dp[idx] > 0 || s.charAt(idx) == '0') return dp[idx];
                                                                    
                                                                             dp[idx] = dfs(s, idx + 1, dp);
                                                                                      if (idx + 1 < s.length() && (s.charAt(idx) == '1' || s.charAt(idx) == '2' && s.charAt(idx + 1) <= '6'))
                                                                                                   dp[idx] += dfs(s, idx + 2, dp);
                                                                                                    
                                                                                                             return dp[idx];
                                                                                                                  }
                                                                                                                   
                                                                                                                    }

File: Data Structures & Algorithms/subsets/submission-1.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []

        def dfs(idx: int):
            if idx == len(nums):
                res.append(tmp[:])  # Make a copy
                return

            # Include nums[idx]
            tmp.append(nums[idx])
            dfs(idx + 1)
            tmp.pop()

            # Exclude nums[idx]
            dfs(idx + 1)

        tmp = []
        dfs(0)
        return res